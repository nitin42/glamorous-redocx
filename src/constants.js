const CHANNEL = '__glamorous__';
export default CHANNEL;
