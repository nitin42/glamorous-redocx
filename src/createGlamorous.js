import React from 'react';
import PropTypes from 'prop-types';
import _ from 'underscore';
import getStyles from './styles';
import CHANNEL from './constants';

/**
 * returns glamorous component factory
 * @param {*} splitProps split the component props
 */
export default function createGlamorous(splitProps) {
  /**
   * glamorous constructor
   * @param {*} comp component name
   * @param {*} param0 display name of the component
   * @param {*} param1 forward props (extra props)
   * @param {*} param2 rootEl (root element)
   * @param {*} param3 propsAreStyleOverrides (built in glamorous component)
   */
  return function glamorous(
    comp,
    /* eslint-disable no-unused-vars */
    { displayName, forwardProps = [], rootEl, propsAreStyleOverrides } = {},
  ) {
    /* eslint-disable no-use-before-define */
    return glamorousComponentFactory;

    /**
     * returns a glamorous component
     * @param {*} styles Redocx styles
     */
    function glamorousComponentFactory(...styles) {
      class GlamorousComponent extends React.Component {
        state = { theme: null };

        /**
         * set the theme
         */
        componentWillMount() {
          /* eslint-disable react/prop-types */
          const { theme } = this.props;

          if (this.context[CHANNEL]) {
            /* eslint-disable no-unneeded-ternary */
            this.setTheme(theme ? theme : this.context[CHANNEL].getState());
          } else {
            this.setTheme(theme || {});
          }
        }

        /**
         * Subscribe to the store which has a theme object
         */
        componentDidMount() {
          if (this.context[CHANNEL] && !this.props.theme) {
            this.unsubscribe = this.context[CHANNEL].subscribe(this.setTheme);
          }
        }

        /**
         * update the new theme
         * @param {*} nextProps new theme
         */
        componentWillReceiveProps(nextProps) {
          if (this.props.theme !== nextProps.theme) {
            this.setTheme(nextProps.theme);
          }
        }

        /**
         * clear all the subscriptions
         */
        componentWillUnmount() {
          /* eslint-disable no-unused-expressions */
          this.unsubscribe;
        }

        /** set the new theme */
        setTheme = theme => this.setState({ theme });

        render() {
          const props = this.props;

          const { styleOverrides, toForward } = splitProps(props, GlamorousComponent);

          /** returns an array of styles */
          const applyStyles = getStyles(styles, props, this.state.theme);

          /** Component props and the extra props to be forwarded */
          const compProps = { ...props, ...toForward };

          return React.createElement(GlamorousComponent.comp, {
            ...compProps,
            style: _.extend.apply(null, [{}].concat(applyStyles, styleOverrides)) || {},
          });
        }
      }

      GlamorousComponent.comp = comp;

      GlamorousComponent.contextTypes = {
        [CHANNEL]: PropTypes.object,
      };

      /**
       * create a new component by extending the styles of old component
       * @param {*} newComp new component
       * @param {*} options options
       */
      function withComponent(newComp, options = {}) {
        return glamorous(comp, {
          forwardProps: GlamorousComponent.forwardProps,
          ...options,
        })(...GlamorousComponent.styles);
      }

      /**
       * assign component data to the glamorous constructor
       */
      Object.assign(
        GlamorousComponent,
        getGlamorousComponentMetadata({
          comp,
          styles,
          displayName,
          forwardProps,
          rootEl,
        }),
        { withComponent },
      );

      return GlamorousComponent;
    }
  };
}

/**
 * returns data about the component
 * @param {*} param0 component
 * @param {*} param1 displayName - display name of the component and component styles
 * @param {*} param2 styles - component styles
 * @param {*} param3 rootEl - represent the root element
 * @param {*} param4 forwardProps - extra props to be forwarded to the component
 */
function getGlamorousComponentMetadata({ comp, displayName, styles, rootEl, forwardProps }) {
  const componentsComp = comp.comp ? comp.comp : comp;

  return {
    comp: componentsComp,
    styles: when(comp.styles, styles),
    rootEl: rootEl || componentsComp,
    forwardProps: when(comp.forwardProps, forwardProps),
    displayName: displayName || `glamorous(${getDisplayName(comp)})`,
  };
}

/**
 * Merge the styles
 * @param {*} comp component
 * @param {*} prop component props
 */
function when(comp, prop) {
  return comp ? comp.concat(prop) : prop;
}

/**
 * returns the display name of the component
 * @param {*} comp component
 */
function getDisplayName(comp) {
  return typeof comp === 'string' ? comp : comp.displayName || comp.name || 'unknown';
}
