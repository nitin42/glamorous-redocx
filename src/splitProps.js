const COMMON_PROPS = ['align', 'style'];

const TextElements = ['TEXT', 'NUMBERITEM', 'BULLETITEM', 'FOOTER', 'HEADER'];
const TextProps = COMMON_PROPS;

const TableElement = 'TABLE';
const TableProps = ['headers', 'data', 'style'];

const ImageElement = 'IMAGE';
const ImageProps = ['width', 'height', 'style', 'align'];

const WrapperElements = 'DOCUMENT';
const WrapperProps = ['align', 'style', 'info'];

const ListWrapper = 'LIST';
const ListProps = COMMON_PROPS;

/**
 * Currently supported style attribute for redocx components
 * @param {*} propName prop name
 */
function isStyleAttribute(propName) {
  const styleAttrs = [
    'fontSize',
    'color',
    'backgroundColor',
    'bold',
    'border',
    'borderColor',
    'fontSize',
    'fontFace',
    'highlight',
    'italic',
    'link',
    'underline',
    'width',
    'height',
    'align',
  ];

  return styleAttrs.includes(propName);
}

/**
 * Check whether it redocx prop or not
 * @param {*} rootEl root element
 * @param {*} prop component prop
 */
function isRedocxProp(rootEl, prop) {
  const REDOCX_ELEMENT = rootEl.toUpperCase();

  if (TextElements.indexOf(REDOCX_ELEMENT) > -1) {
    return TextProps.indexOf(prop) > -1;
  }

  if (TableElement === REDOCX_ELEMENT) {
    return TableProps.indexOf(prop) > -1;
  }

  if (ImageElement === REDOCX_ELEMENT) {
    return ImageProps.indexOf(prop) > -1;
  }

  if (WrapperElements === REDOCX_ELEMENT) {
    return WrapperProps.indexOf(prop) > -1;
  }

  if (ListWrapper === REDOCX_ELEMENT) {
    return ListProps.indexOf(prop) > -1;
  }

  return false;
}

/**
 * Check whether the property can be passed to a component
 * @param {*} rootEl root element
 * @param {*} prop prop name
 */
function shouldForwardProperty(rootEl, prop) {
  return typeof rootEl !== 'string' && !isRedocxProp(rootEl, prop);
}

/**
 * Split the component props
 * @param {*} props Component props
 * @param {*} component Component
 */
export default function splitProps(props, component) {
  const { rootEl, forwardProps } = component;

  const returnValue = { styleOverrides: {}, toForward: {} };

  return Object.keys(props).reduce((container, prop) => {
    if (forwardProps.indexOf(prop) !== -1 || shouldForwardProperty(rootEl, prop)) {
      /* eslint-disable  no-param-reassign */
      container.toForward[prop] = props[prop];
    } else if (isStyleAttribute(prop)) {
      /* eslint-disable  no-param-reassign */
      container.styleOverrides[prop] = props[prop];
    }
    return container;
  }, returnValue);
}
