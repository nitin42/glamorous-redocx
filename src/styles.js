/**
 * Evaluate the function with props or return style if an object
 * @param {*} styles Component styles
 * @param {*} props Component props
 */
function evaluateGlamorStyles(styles, props, theme) {
  return styles.map((style) => {
    if (typeof style === 'function') {
      return style(props, theme);
    }
    return style;
  });
}

/**
 * Get the styles and assign them to the 'style' prop of the component
 * @param {*} styles Component styles
 * @param {*} props Component props
 */
function getStyles(styles, props, theme) {
  const glamorStyles = evaluateGlamorStyles(styles, props, theme);
  const outputStyles = glamorStyles;

  if (props.style) {
    outputStyles.push(props.style);
  }

  return outputStyles;
}

export default getStyles;
