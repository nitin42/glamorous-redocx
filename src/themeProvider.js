import React from 'react';
import brcast from 'brcast';
import PropTypes from 'prop-types';
import CHANNEL from './constants';

export default class ThemeProvider extends React.Component {
  static childContextTypes = {
    [CHANNEL]: PropTypes.object.isRequired,
  };

  static contextTypes = {
    /* eslint-disable react/forbid-prop-types */
    [CHANNEL]: PropTypes.object,
  };

  static propTypes = {
    theme: PropTypes.object.isRequired,
    /* eslint-disable react/require-default-props */
    children: PropTypes.node,
  };

  getChildContext() {
    return {
      [CHANNEL]: this.broadcast,
    };
  }

  componentWillMount() {
    if (this.context[CHANNEL]) {
      this.setOuterTheme(this.context[CHANNEL].getState());
      this.broadcast.setState(this.getTheme());
    }
  }

  componentDidMount() {
    if (this.context[CHANNEL]) {
      this.unsubscribe = this.context[CHANNEL].subscribe(this.setOuterTheme);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.theme !== nextProps.theme) {
      this.broadcast.setState(this.getTheme(nextProps.theme));
    }
  }

  componentWillUnmount() {
    /* eslint-disable no-unused-expressions */
    this.unsubscribe && this.unsubscribe();
  }

  getTheme(passedTheme) {
    const theme = passedTheme || this.props.theme;
    return { ...this.outerTheme, ...theme };
  }

  setOuterTheme = (theme) => {
    this.outerTheme = theme;
  }

  broadcast = brcast(this.props.theme);

  render() {
    return this.props.children ? React.Children.only(this.props.children) : null;
  }
}
