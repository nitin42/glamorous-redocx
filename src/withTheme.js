import React from 'react';
import PropTypes from 'prop-types';
import CHANNEL from './constants';

export default function withTheme(ComponentToTheme) {
  class ThemedComponent extends React.Component {
    state = { theme: {} };

    componentWillMount() {
      this.setState({ theme: this.context[CHANNEL].getState() });
    }

    componentDidMount() {
      if (this.context[CHANNEL]) {
        this.unsubscribe = this.context[CHANNEL].subscribe(this.setTheme);
      }
    }

    componentWillUnmount() {
      /* eslint-disable no-unused-expressions */
      this.unsubscribe;
    }

    setTheme = theme => this.setState({ theme });

    render() {
      /* eslint-disable react/jsx-filename-extension */
      return <ComponentToTheme {...this.props} {...this.state} />;
    }
  }

  ThemedComponent.contextTypes = {
    [CHANNEL]: PropTypes.object,
  };

  return ThemedComponent;
}
