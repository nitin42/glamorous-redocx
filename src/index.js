import createGlamorous from './createGlamorous';
import splitProps from './splitProps';
import ThemeProvider from './themeProvider';
import withTheme from './withTheme';

const glamorous = createGlamorous(splitProps);

const RedocxElements = ['Text', 'Document', 'Image', 'Header', 'Footer', 'List', 'NumberItem', 'BulletItem', 'Table'];

Object.assign(
  glamorous,
  RedocxElements.reduce((getters, tag) => {
    const docTag = tag.toUpperCase();
    /* eslint-disable no-param-reassign */
    getters[tag] = glamorous(docTag);
    return getters;
  }, {}),
);

Object.assign(
  glamorous,
  RedocxElements.reduce((container, tag) => {
    const docTag = tag.toUpperCase();
    container[docTag] = glamorous[tag]();
    container[docTag].displayName = `glamorous.${tag}`;
    container[docTag].propsAreStyleOverrides = true;
    return container;
  }, {}),
);

export default glamorous;

export { ThemeProvider, withTheme };
