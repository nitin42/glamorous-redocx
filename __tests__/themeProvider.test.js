import React, {Component} from 'react'
import {render, mount} from 'enzyme'
import serializer from 'jest-glamor-react'
import glamorous from '../src/'
import ThemeProvider from '../src/themeProvider'
import {CHANNEL} from '../src/constants'

expect.addSnapshotSerializer(serializer)

const getMockedContext = unsubscribe => ({
  [CHANNEL]: {
    getState: () => {},
    setState: () => {},
    subscribe: () => unsubscribe,
  },
})

test('renders a component with theme', () => {
  const Comp = glamorous.Text(
    {
      fontSize: 40,
    },
    (props, theme) => ({color: theme.color}),
  )
  expect(
    render(
      <ThemeProvider theme={{color: 'red'}}>
        <Comp />
      </ThemeProvider>,
    ),
  ).toMatchSnapshot()
})

test('theme properties updates get propagated down the tree', () => {
  class Parent extends Component {
    state = {
      fontSize: 10,
    }

    render() {
      return (
        <ThemeProvider theme={{fontSize: this.state.fontSize}}>
          <Child />
        </ThemeProvider>
      )
    }
  }
  const Child = glamorous.Text(
    {
      color: 'red',
    },
    (props, theme) => ({fontSize: theme.fontSize}),
  )
  const wrapper = mount(<Parent />)
  expect(wrapper).toMatchSnapshot(`with theme prop of fontSize 10`)
  wrapper.setState({fontSize: 20})
  expect(wrapper).toMatchSnapshot(`with theme prop of fontSize 20`)
})

test('merges nested themes', () => {
  const One = glamorous.Text({}, (props, {fontSize, bold}) => ({
    fontSize,
    bold,
  }))
  const Two = glamorous.Text({}, (props, {fontSize, bold}) => ({
    fontSize,
    bold,
  }))
  expect(
    mount(
      <glamorous.DOCUMENT>
        <ThemeProvider theme={{fontSize: 10, bold: true}}>
          <glamorous.DOCUMENT>
            <One />
            <ThemeProvider theme={{fontSize: 20}}>
              <Two />
            </ThemeProvider>
          </glamorous.DOCUMENT>
        </ThemeProvider>
      </glamorous.DOCUMENT>,
    ),
  ).toMatchSnapshot()
})

test('renders if children are null', () => {
  expect(
    mount(
      <ThemeProvider theme={{fontSize: 10}}>
        {false && <glamorous.TEXT />}
      </ThemeProvider>,
    ),
  ).toMatchSnapshot()
})

test('does nothing when receive same theme via props', () => {
  const theme = {fontSize: 20}
  const wrapper = mount(<ThemeProvider theme={theme} />)
  expect(wrapper).toMatchSnapshot(`with theme prop of fontSize 20`)
  wrapper.setProps({theme})
  expect(wrapper).toMatchSnapshot(`with theme prop of fontSize 20`)
})