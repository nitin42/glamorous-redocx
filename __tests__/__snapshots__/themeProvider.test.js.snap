// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merges nested themes 1`] = `
<glamorous.Document>
  <DOCUMENT
    style={Object {}}
  >
    <ThemeProvider
      theme={
        Object {
          "bold": true,
          "fontSize": 10,
        }
      }
    >
      <glamorous.Document>
        <DOCUMENT
          style={Object {}}
        >
          <glamorous(TEXT)>
            <TEXT
              style={
                Object {
                  "bold": true,
                  "fontSize": 10,
                }
              }
            />
          </glamorous(TEXT)>
          <ThemeProvider
            theme={
              Object {
                "fontSize": 20,
              }
            }
          >
            <glamorous(TEXT)>
              <TEXT
                style={
                  Object {
                    "bold": true,
                    "fontSize": 20,
                  }
                }
              />
            </glamorous(TEXT)>
          </ThemeProvider>
        </DOCUMENT>
      </glamorous.Document>
    </ThemeProvider>
  </DOCUMENT>
</glamorous.Document>
`;

exports[`renders a component with theme 1`] = `
<text
  style="font-size:40px;color:red;"
/>
`;

exports[`renders if children are null 1`] = `
<ThemeProvider
  theme={
    Object {
      "fontSize": 10,
    }
  }
/>
`;

exports[`with theme prop of fontSize 10 1`] = `
<Parent>
  <ThemeProvider
    theme={
      Object {
        "fontSize": 10,
      }
    }
  >
    <glamorous(TEXT)>
      <TEXT
        style={
          Object {
            "color": "red",
            "fontSize": 10,
          }
        }
      />
    </glamorous(TEXT)>
  </ThemeProvider>
</Parent>
`;

exports[`with theme prop of fontSize 20 1`] = `
<Parent>
  <ThemeProvider
    theme={
      Object {
        "fontSize": 20,
      }
    }
  >
    <glamorous(TEXT)>
      <TEXT
        style={
          Object {
            "color": "red",
            "fontSize": 20,
          }
        }
      />
    </glamorous(TEXT)>
  </ThemeProvider>
</Parent>
`;

exports[`with theme prop of fontSize 20 2`] = `
<ThemeProvider
  theme={
    Object {
      "fontSize": 20,
    }
  }
/>
`;

exports[`with theme prop of fontSize 20 3`] = `
<ThemeProvider
  theme={
    Object {
      "fontSize": 20,
    }
  }
/>
`;
